cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/rolling/RackRolling.java

```diff
@@ -80,6 +86,55 @@ private static boolean avail(KafkaNode kafkaNode,
         return true;
     }
 
+    private static boolean isQuorumHealthyWithoutNode(Reconciliation reconciliation, int controllerNeedRestarting, int activeControllerId, Map<Integer, Long> quorumFollowerStates) {
+        LOGGER.debugCr(reconciliation, "Determining the impact of restarting controller {} on quorum health", controllerNeedRestarting);
+        if (activeControllerId < 0) {
+            LOGGER.warnCr(reconciliation, "No controller quorum leader is found because the leader id is set to {}", activeControllerId);
+            return false;
+        }
+
+        int totalNumOfControllers = quorumFollowerStates.size();
+        if (totalNumOfControllers == 1) {
+            LOGGER.warnCr(reconciliation, "Performing rolling update on a controller quorum with a single node. The cluster may be " +
+                    "in a defective state once the rolling update is complete. It is recommended that a minimum of three controllers are used.");
+            return true;
+        }
+
+        long leaderLastCaughtUpTimestamp = quorumFollowerStates.get(activeControllerId);
+
+        long numOfCaughtUpControllers = quorumFollowerStates.entrySet().stream().filter(entry -> {
+            int nodeId = entry.getKey();
+            long lastCaughtUpTimestamp = entry.getValue();
+            if (lastCaughtUpTimestamp < 0) {
+                LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for controller {} ", nodeId);
+            } else {
+                LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for controller {} is {}", nodeId, lastCaughtUpTimestamp);
+                if (nodeId == activeControllerId || (leaderLastCaughtUpTimestamp - lastCaughtUpTimestamp) < controllerQuorumFetchTimeout) {
+                    if (nodeId != controllerNeedRestarting) {
+                        return true;
+                    }
+                    LOGGER.debugCr(reconciliation, "Controller {} has caught up with the controller quorum leader", nodeId);
+                } else {
+                    LOGGER.debugCr(reconciliation, "Controller {} has fallen behind the controller quorum leader", nodeId);
+                }
+            }
+            return false;
+        }).count();
+
+        if (totalNumOfControllers == 2) {
+            // Only roll the controller if the other one in the quorum has caught up or is the active controller.
+            if (numOfCaughtUpControllers == 1) {
+                LOGGER.warnCr(reconciliation, "Performing rolling update on a controller quorum with 2 nodes. The cluster may be " +
+                        "in a defective state once the rolling update is complete. It is recommended that a minimum of three controllers are used.");
+                return true;
+            } else {
+                return false;
+            }
+        } else {
+            return numOfCaughtUpControllers >= ceil((double) (totalNumOfControllers + 1) / 2);
```

Comment:
Didn't we managed to avoid `double` arithmetic last time?

Reply:
you are right, we did. I had done this change before that.